### ISSUE TITLE
Create types based on Key

### ISSUE THREAD

**Issue by julianhyde** [hydromatic/morel#43]
Create types based on Key.

Fixing this should help us model packages (e.g. `List`) as if they were records. Currently we have trouble when the fields in a record have polymorphic types (e.g. `List.hd` has type `'a list -> 'a`) and the type variables get confused when we extract a single field from a record type.

**Comment by julianhyde** [comment]
As part of the fix, enable the following cases in `simple.sml`, which are commented out because they are broken:
```
val SOME i = SOME 1;
val SOME (i, NONE) = SOME (1, SOME false);
```

**Comment by julianhyde** [comment]
Fixed as part of #70.
